### PR TITLE
refactor(storage): separate Read image values from SQLite

### PR DESCRIPTION
**Pull request by likun666661**
## Summary

- Add an explicit `inline | managed_file` value representation to context offload. Small text/JSON values can remain inline; binary and intermediate-file owners can select managed-file storage.
- Upgrade the context-offload schema to v3. For managed values, `context_blobs.payload` contains a short root-bound content-addressed locator such as `sha256/ab/<digest>` while `size_bytes`, quotas, deduplication, and identity continue to describe the actual content.
- Publish managed values under the authenticated Storage Root with a durable SQLite deletion intent before file publication, bounded stable reads, and SHA-256 verification. SQLite never accepts an arbitrary absolute path.
- Keep deletion metadata in the indexed `context_file_deletions` queue. GC deletes database ownership transactionally, drains exact locators without a global directory scan, and cancels stale deletion work when the same managed value is reused.
- Make `ReadImageSnapshotStore` always select `managed_file`, retain its 5 MiB product cap, stable `(Session, ownerId)` identity, Session-authorized reads, and fail-closed owner/media validation.
- Migrate existing v1/v2 context payloads as inline values; no eager filesystem rewrite is required.

This is the value-separation correction to the typed Read image slice. The data path is now:

`SessionContextRef -> SQLite ref/value metadata -> locator BLOB -> managed intermediate file`

This slice still does not add `SessionContextRef` to `StorageRef`, wire builtin Read, publish a new storage package entrypoint, or activate new production writes.

Refs #4071

## Verification

- `npm --workspace @maka/core run build`
- `npm --workspace @maka/storage run build`
- Targeted context-offload and Read image suites: 28/28 pass
- `NODE_NO_WARNINGS=1 node --test --test-reporter=dot "packages/storage/dist/**/*.test.js"` — full storage suite passes (1037 tests: 1021 pass, 16 skip)
- `npx biome check` on all changed TypeScript files
- `npm run check:asf-headers`

## AI use

- [ ] No generative tool made a substantive contribution
- [x] Generative tooling made a substantive contribution

Tool and scope: OpenAI Codex contributed repository inspection, value-separation design, implementation, tests, and PR text. The commit includes the required `Generated-by` trailer.

## Checklist

- [x] Tests cover the change and fail without it
- [x] Lint, format, typecheck, affected suites, and the full storage suite pass locally

Does this PR entail a change in behavior?

- [ ] Yes
- [x] No — the typed facade remains unwired to production consumers

### CODE REVIEW

File: packages/storage/src/sqlite-context-offload-store.ts [hunk 1]
```diff
@@ -317,64 +372,91 @@ export class SqliteContextOffloadStore implements ContextOffloadStore {
...
-        hasMore: rows.length > selected.length,
-      };
+            )
+            .run(selected.length, deletedBytes);
+          if (updated.changes !== 1) throw new Error('Missing context store usage row');
+        }
+        return {
+          deletedBlobs: selected.length,
+          deletedBytes,
+          hasMore: rows.length > selected.length,
```
Comment (Astro-Han) [inline packages/storage/src/sqlite-context-offload-store.ts]: **[bug]** `hasMore` is computed only from the blob-candidate SELECT (`rows.length > selected.length`) before the second file-deletion drain. Each `collectGarbage` call drains at most `2 * maxBlobs` `context_file_deletions` rows (once at line 377, once at line 458). Crash orphans live only in that queue: they are not blob candidates, so a call can leave files on disk and still return `{ deletedBlobs: 0, hasMore: false }`. Reachable path: three managed puts that crash after `after_managed_file_publish`, then `collectGarbage({ maxBlobs: 1 })` — two locators are drained, one file remains, and a `while (result.hasMore)` caller never continues. The crash-recovery test uses a single orphan, so it cannot catch this, even though deletion-intent recovery is the point of that test.

**Suggestion:** After the post-txn drain, set `hasMore` true if any `context_file_deletions` rows remain (or include `remainingDeletions > 0` in the same flag). Add a test with three crash orphans and `maxBlobs: 1` that asserts `hasMore: true` until the queue is empty and every file is gone.

File: packages/storage/src/sqlite-context-offload-store.ts [hunk 2]
```diff
@@ -695,11 +830,215 @@ export class SqliteContextOffloadStore implements ContextOffloadStore {
...
+      await handle.close();
+      handle = undefined;
+      try {
+        await link(temporary, target);
+        await this.#assertManagedDirectory(targetDirectory);
+        await syncDirectoryChain(targetDirectory, storageRoot);
+      } catch (error) {
+        if (!isNodeError(error, 'EEXIST')) throw error;
+        await this.#verifyManagedFile(target, blobId, bytes.byteLength);
+      }
```
Comment (Astro-Han) [inline packages/storage/src/sqlite-context-offload-store.ts]: **[suggestion]** The create path fsyncs the temp file and then `syncDirectoryChain` after `link`. The `EEXIST` reuse path only `#verifyManagedFile` and returns. That is the path taken when a crashed publisher left a file behind and a later put adopts it. If the first process died after `link` and before the directory fsync, a reuse can commit a live `context_blobs` row whose dirent is still not durable; a subsequent power loss drops the file and reads return `corrupt`.

**Suggestion:** On `EEXIST`, after a successful verify, fsync the existing file and run the same `syncDirectoryChain(targetDirectory, storageRoot)` before returning the locator. Cover it by crashing after `link` and before directory sync, then putting the same bytes on reopen.

File: packages/storage/src/sqlite-context-offload-store.ts (hunk 1 above)
Comment (Astro-Han) [inline packages/storage/src/sqlite-context-offload-store.ts:482]: `hasMore` is still `rows.length > selected.length` from the blob-candidate SELECT, and it's captured before the second drain. Crash orphans never appear in that query, so they don't count.

One leftover file is fine; that's the crash-child case. Three distinct crashed publishes plus `collectGarbage({ maxBlobs: 1 })` drains two locators and still returns `{ deletedBlobs: 0, hasMore: false }` with a file on disk. A `while (result.hasMore)` caller stops. A later independent GC call would pick it up, so I'm not blocking on it.

If `hasMore` is supposed to mean "this store still has cleanup work," leftover `context_file_deletions` should flip it after the post-txn drain. A three-orphan / `maxBlobs: 1` test would make that stick. If you meant blob candidates only, a sentence on the result type would help, because the deletion queue is otherwise invisible.

File: packages/storage/src/sqlite-context-offload-store.ts (hunk 2 above)
Comment (Astro-Han) [inline packages/storage/src/sqlite-context-offload-store.ts:947]: Create fsyncs the temp file, then `syncDirectoryChain` after `link`. The `EEXIST` branch just verifies bytes and returns.

That's how you adopt a file left by a crash. If the first process died after `link` and before the directory fsync, a retry of the same bytes commits a live blob whose dirent still isn't durable. A later power loss can drop the file, and reads come back `corrupt`.

Same durability rule as create, I think: after a successful verify, fsync the existing file and run `syncDirectoryChain(targetDirectory, storageRoot)` before returning.

**Review by Astro-Han — APPROVED** [review]

**Review by M4n5ter — COMMENTED** [review]
<details open>
<summary>English</summary>

I don't think this shape is ready to merge. The typed Read boundary is useful, but the new managed-file substrate currently has correctness gaps, and it reverses the SQLite-first decision in #4071 without the evidence that issue required.

### Blocking findings

1. **The Read facade authorizes a reference using the Session declared by that same reference.**

   [`ReadImageSnapshotStore.read()`](https://github.com/apache/maka/blob/01de2ccfc19684f1a5e7b063c82bad4b8a9d6dda/packages/storage/src/read-image-snapshot-store.ts#L75-L80) forwards `input.sessionId` directly to the physical store. A caller in Session B that possesses Session A's complete `SessionContextRef` supplies A as both the claimed and checked Session, so the read succeeds. This does not satisfy #4071's invariant that cross-Session reads fail even when the caller knows the ref identity. The facade should be Session-scoped, or receive the authenticated current Session separately and compare it with the ref before reading.

2. **Per-call `storageKind` is incompatible with blob-level deduplication and can acknowledge an unreadable value.**

   [`#promoteToManagedFile()`](https://github.com/apache/maka/blob/01de2ccfc19684f1a5e7b063c82bad4b8a9d6dda/packages/storage/src/sqlite-context-offload-store.ts#L617-L633) promotes the one shared blob row from inline to managed when any owner selects `managed_file`. [`blobMatchesInput()`](https://github.com/apache/maka/blob/01de2ccfc19684f1a5e7b063c82bad4b8a9d6dda/packages/storage/src/sqlite-context-offload-store.ts#L1163-L1175) then treats an existing managed blob as matching without verifying its file.

   I reproduced this sequence:

   - store bytes for a Tool Result as inline;
   - store identical bytes for a Read owner as managed;
   - delete the managed file;
   - retry the inline Tool Result put with the original bytes.

   The retry returns `ok: true`, but reading that same ref immediately returns `corrupt`. The root issue is that the producer appears to choose a representation per value occurrence, while the implementation stores one mutable representation per SHA-256 blob. Physical tiering should be an authority-owned global policy; if representations must remain independently selectable, representation has to be part of the physical identity.

3. **The `EEXIST` publication recovery path can commit a reference to a non-durable directory entry.**

   In [`#publishManagedFile()`](https://github.com/apache/maka/blob/01de2ccfc19684f1a5e7b063c82bad4b8a9d6dda/packages/storage/src/sqlite-context-offload-store.ts#L879-L886), `syncDirectoryChain()` runs only after a newly created hard link. If process A exits after `link()` but before the directory sync, a retry can observe `EEXIST`, verify the bytes, commit the SQLite reference, and cancel the deletion intent without ever establishing directory durability. A later machine crash can lose the locator while retaining the database reference. Both the new-link and verified-`EEXIST` paths need the same directory durability barrier before the database commit.

4. **Publication temporary files are outside every recovery and quota authority.**

   [`#publishManagedFile()`](https://github.com/apache/maka/blob/01de2ccfc19684f1a5e7b063c82bad4b8a9d6dda/packages/storage/src/sqlite-context-offload-store.ts#L871-L892) creates a random full-size `.tmp`, but the durable intent and recovery queue only name the final locator. A process exit during publication permanently leaks the temporary file. On the successful-link path, the directory is synced while the temporary hard link still exists, then the temp is unlinked without syncing that deletion, so a later crash can also restore it. These files are invisible to `usage()`, workspace quota, Session retirement, and GC. Staging needs a recoverable identity, or the exact staging name must participate in the durable intent and cleanup protocol.

### Important finding

5. **GC reports completion and releases physical quota before file deletion is complete.**

   [`context_store_usage.physical_bytes` is decremented](https://github.com/apache/maka/blob/01de2ccfc19684f1a5e7b063c82bad4b8a9d6dda/packages/storage/src/sqlite-context-offload-store.ts#L442-L450) before the managed file is durably removed, while [`hasMore`](https://github.com/apache/maka/blob/01de2ccfc19684f1a5e7b063c82bad4b8a9d6dda/packages/storage/src/sqlite-context-offload-store.ts#L452-L459) only reflects blob candidates and ignores remaining `context_file_deletions`. After unlink failure or a DB-commit/process-exit window, new puts can consume quota already released for bytes still occupying the Storage Root. With several publication intents, GC can also return `hasMore: false` while recovery work remains. Pending deletion bytes and continuation state need to remain part of accounting until durable deletion succeeds.

### Architectural direction

The sharper concern is that this PR reintroduces exactly the SQLite + filesystem durability split that #4071 deliberately deferred. The accepted issue says the current consumers are capped whole-object values and selects monolithic SQLite first; external CAS was gated on measured large-object, streaming, or performance evidence. This PR adds schema v3, publication staging, deletion intents, locator validation, stable filesystem reads, and another recovery lifecycle without presenting that evidence.

I would make the current slice smaller and cleaner:

- keep the typed, Session-scoped `ReadImageSnapshotStore`;
- back it with the existing single-domain SQLite BLOB authority for now;
- remove the caller-facing `ContextOffloadStorageKind` contract;
- do not freeze schema v3 and managed CAS until 5 MiB BLOB measurements show that SQLite misses an explicit event-loop/WAL/checkpoint/backup budget, or a real consumer requires path/streaming/range semantics.

At the larger boundary, every model-visible value should have one canonical byte owner. Read image snapshots naturally need a context owner because the workspace source can change. Tool Results currently still have their complete body in the immutable RuntimeEvent ledger, so the archive remains a second materialization until the T2 contract is changed to persist only a durable context ref for large results. That canonicality decision should come before making Read images and Tool Results share a physical tiering model.

### Verification

- `@maka/core` and `@maka/storage` builds passed.
- The three targeted suites passed, 28/28.
- The full storage suite exited successfully.
- A separate directed probe reproduced the inline-put-success / managed-read-corrupt sequence above.

Correctness verdict: **not acceptable**.  
Design verdict: **not acceptable**.

</details>

<details>
<summary>中文</summary>

我认为当前形态还不能合并。typed Read 边界本身有价值，但新 managed-file 底座存在 correctness 缺口，而且它在没有提供 #4071 所要求证据的情况下，反转了已经确定的 SQLite-first 决策。

### Blocking findings

1. **Read facade 使用引用自己声明的 Session 来授权该引用。**

   [`ReadImageSnapshotStore.read()`](https://github.com/apache/maka/blob/01de2ccfc19684f1a5e7b063c82bad4b8a9d6dda/packages/storage/src/read-image-snapshot-store.ts#L75-L80) 直接把 `input.sessionId` 传给底层。Session B 的调用者只要持有 Session A 的完整 `SessionContextRef`，就会同时以 A 作为声明和校验 Session，读取会成功。这不满足 #4071 中“即使调用方知道 ref identity，跨 Session 读取也必须失败”的不变量。facade 应绑定到经过认证的当前 Session，或者单独接收可信的当前 Session，并在读取前与 ref 比较。

2. **per-call `storageKind` 与 blob-level 去重不相容，并且能够对不可读值返回成功。**

   [`#promoteToManagedFile()`](https://github.com/apache/maka/blob/01de2ccfc19684f1a5e7b063c82bad4b8a9d6dda/packages/storage/src/sqlite-context-offload-store.ts#L617-L633) 会在任意 owner 选择 `managed_file` 时，把唯一共享的 blob 行从 inline 全局提升为 managed；随后 [`blobMatchesInput()`](https://github.com/apache/maka/blob/01de2ccfc19684f1a5e7b063c82bad4b8a9d6dda/packages/storage/src/sqlite-context-offload-store.ts#L1163-L1175) 对已有 managed blob 不验证实际文件便认为匹配。

   我复现了下面的路径：

   - Tool Result 以 inline 保存一组字节；
   - Read owner 以 managed 保存完全相同的字节；
   - 删除 managed 文件；
   - 用原始字节重试 inline Tool Result put。

   重试返回 `ok: true`，但同一 ref 随即读取却返回 `corrupt`。根因是 API 看起来允许 producer 为每个 value occurrence 选择 representation，而实现实际上为每个 SHA-256 blob 只保存一个可变 representation。物理 tier 应由 authority 以全局策略决定；如果表示必须可以独立选择，就必须把 representation 纳入物理 identity。

3. **`EEXIST` publication 恢复路径可能提交指向非持久目录项的引用。**

   [`#publishManagedFile()`](https://github.com/apache/maka/blob/01de2ccfc19684f1a5e7b063c82bad4b8a9d6dda/packages/storage/src/sqlite-context-offload-store.ts#L879-L886) 只在新建 hard link 后执行 `syncDirectoryChain()`。如果进程 A 在 `link()` 后、目录同步前退出，重试可能看到 `EEXIST`、校验字节、提交 SQLite ref 并清除删除 intent，却从未建立目录持久性。之后机器崩溃可能丢失 locator，但数据库 ref 仍然存在。新建 link 与校验 `EEXIST` 两条路径都必须在 DB commit 前经过相同的目录 durability barrier。

4. **publication 临时文件不属于任何 recovery 或 quota authority。**

   [`#publishManagedFile()`](https://github.com/apache/maka/blob/01de2ccfc19684f1a5e7b063c82bad4b8a9d6dda/packages/storage/src/sqlite-context-offload-store.ts#L871-L892) 创建随机、完整大小的 `.tmp`，但 durable intent 和恢复队列只记录最终 locator。进程在 publication 期间退出会永久泄漏临时文件。成功建 link 时，代码在 temp hardlink 仍存在时同步目录，随后 unlink temp 却不持久化这次删除，因此后续崩溃也可能让它重新出现。这些文件不受 `usage()`、workspace quota、Session retirement 或 GC 管理。staging 必须拥有可恢复 identity，或者精确 staging 名称必须进入 durable intent 与清理协议。

### Important finding

5. **GC 在文件删除完成前就报告完成并释放物理配额。**

   [`context_store_usage.physical_bytes` 在 managed 文件持久删除前已经扣减](https://github.com/apache/maka/blob/01de2ccfc19684f1a5e7b063c82bad4b8a9d6dda/packages/storage/src/sqlite-context-offload-store.ts#L442-L450)，而 [`hasMore`](https://github.com/apache/maka/blob/01de2ccfc19684f1a5e7b063c82bad4b8a9d6dda/packages/storage/src/sqlite-context-offload-store.ts#L452-L459) 只反映 blob candidates，不包含剩余的 `context_file_deletions`。unlink 失败，或进程在 DB commit 后退出时，新 put 可以使用已经释放、但仍由 Storage Root 实际占用的配额。存在多个 publication intent 时，GC 还可能在恢复工作尚未完成时返回 `hasMore: false`。pending deletion bytes 与 continuation state 必须持续进入 accounting，直到持久删除成功。

### 架构方向

更大的问题是：这个 PR 重新引入了 #4071 明确推迟的 SQLite + filesystem 两套 durability domain。该 issue 已确认当前 consumer 是 capped whole-object，并选择先使用 monolithic SQLite；只有 measured large-object、streaming 或性能证据证明不足时，才引入 external CAS。当前 PR 在没有这些证据的情况下增加了 schema v3、publication staging、deletion intent、locator 校验、稳定文件读取和另一套恢复生命周期。

我建议把当前 slice 收小：

- 保留 typed、Session-scoped `ReadImageSnapshotStore`；
- 暂时复用已有单域 SQLite BLOB authority；
- 删除 caller-facing `ContextOffloadStorageKind` 合同；
- 在 5 MiB BLOB 的测量证明 SQLite 无法满足明确的 event-loop/WAL/checkpoint/backup 预算，或者真实 consumer 需要 path/streaming/range semantics 之前，不冻结 schema v3 与 managed CAS。

从更大的边界看，每个 model-visible value 应只有一个 canonical byte owner。Read image snapshot 因 workspace 源文件会变化，天然需要 context owner。Tool Result 的完整 body 当前仍保存在 immutable RuntimeEvent ledger 中，因此在 T2 合同改成大型结果只持久化 context ref 之前，archive 仍然是第二份 materialization。应当先完成 canonicality 决策，再决定 Read image 与 Tool Result 是否共享物理 tier 模型。

### 验证

- `@maka/core` 与 `@maka/storage` build 通过。
- 三组 targeted suites 28/28 通过。
- 完整 storage suite 成功退出。
- 额外定向 probe 稳定复现了上述 inline put 成功、managed read 返回 corrupt 的路径。

Correctness verdict：**not acceptable**。  
Design verdict：**not acceptable**。

</details>

**Comment by likun666661** [comment]
Thanks for the concrete correctness review. I agree with the five implementation findings; they are fixed in `c10089339` rather than used as reasons to hide the value layer:

- `ReadImageSnapshotStore` is now bound to an authenticated Session. A Session-B facade cannot read a complete Session-A ref.
- `storageKind` is removed from the public contract. Placement is authority-owned: Read image values require `managed_file`, Tool Result values currently prefer inline, and a blob already promoted to managed never gets acknowledged through an unchecked inline retry. The exact reproduction above now republishes/verifies the missing file before returning success.
- Both the new-link and verified-`EEXIST` paths establish file + directory durability before the SQLite commit.
- Publication staging has a deterministic content identity. The durable SQLite intent covers both the final locator and its staging name; recovery removes either without enumerating the payload tree, and staging unlink is directory-synced.
- A publication/deletion intent owns physical-byte accounting. GC does not release managed bytes until durable unlink succeeds, and `hasMore` includes remaining deletion intents. A three-intent, `maxBlobs: 1` crash-recovery test now returns `true`, `true`, `false`.

The directed context-offload suites pass 31/31; the full storage suite passes 1040 tests (16 skipped) with the repository's SQLite experimental warning suppressed for the existing warning-sensitive child-process test.

On the architecture conclusion, however, I disagree.

You are right that #4071 currently selects a monolithic SQLite BLOB substrate. This PR is intentionally proposing to amend that physical-placement decision; it is not doing so accidentally. The distinction I think #4071 currently collapses is:

1. **SQLite as the metadata/index/lifecycle authority**, and
2. **SQLite pages as the canonical container for every payload byte**.

We need (1). It does not imply (2).

A 4–5 MiB per-value cap is an API/materialization bound. It is not a bound on the aggregate database, WAL, checkpoint, backup, vacuum, or write-amplification working set. Putting immutable image/PDF/intermediate-file bodies into BLOB pages makes those byte streams participate in every SQLite storage lifecycle merely because their metadata needs transactions. Chunk rows change the page layout; they do not remove that amplification and add a chunk protocol on top.

The hybrid design keeps one control-plane authority:

- SQLite owns refs, content identity, placement, quotas, GC eligibility, deletion intents, and recovery progress.
- The managed file is the value, addressed by a locator stored in SQLite. It owns no searchable metadata and grants no authorization.
- Every cleanup/recovery operation is driven by an indexed SQLite row and an exact locator. There is no `O(all records/files)` reconciliation path and no filesystem scan before Ready.

That is ordinary KV value separation, not a second metadata database or an S3 service. Yes, value separation requires a publication/deletion protocol. The review found real holes in that protocol, and they needed fixing. But the existence of those obligations is not evidence that multi-gigabyte immutable values belong in SQLite's WAL; it is the normal cost of keeping large values out of the LSM/B-tree control plane.

Benchmarks are still useful for selecting an inline threshold, worker placement, and maintenance budgets. I do not think the default burden should be “first prove that routing multi-gigabyte immutable payload populations through SQLite BLOB/WAL/backup is slow enough.” We should instead choose byte ownership from access and lifecycle semantics, then benchmark the thresholds. Images, PDFs, and large intermediate files have file/value-log semantics; small text/JSON has inline-row semantics.

The canonical Tool Result question remains valid but orthogonal. A later T2 slice should make large Tool Results context-canonical and persist only the durable ref; small results can remain ledger-inline. Duplicating a large result into a monolithic SQLite BLOB does not solve canonicality—it only changes where the second copy lives.

So my proposed direction remains: keep SQLite-first for authority and indexing, keep value separation for large binary/intermediate values, and make the crash protocol correct. I would not revert the architecture to monolithic SQLite BLOB storage.

File: packages/storage/src/sqlite-context-offload-store.ts (hunk 2 above)
Comment (likun666661) [reply packages/storage/src/sqlite-context-offload-store.ts:947]: Fixed in c10089339. The verified-EEXIST path now fsyncs the existing value and runs the same directory-chain durability barrier before the SQLite commit.

File: packages/storage/src/sqlite-context-offload-store.ts (hunk 1 above)
Comment (likun666661) [reply packages/storage/src/sqlite-context-offload-store.ts:482]: Fixed in c10089339. GC now treats pending file deletions as bounded continuation work; the three-intent/maxBlobs=1 crash test returns hasMore true, true, then false.

**Review by M4n5ter — APPROVED** [review]
<details open>
<summary>English</summary>

Thanks for the thorough response and the concrete follow-up. I re-reviewed `c10089339` and reran the affected builds and suites.

The five findings are resolved:

- the Read facade is Session-bound;
- physical placement is authority-owned, and the original mixed-representation reproduction now republishes and verifies the missing managed value before acknowledging success;
- both publication paths establish the required durability barrier;
- deterministic staging is covered by the durable intent and recovery path;
- pending file bytes and deletion continuation remain accounted for until durable cleanup completes.

`@maka/core` and `@maka/storage` build successfully, and the directed suites pass 31/31. I did not find a new merge blocker in the follow-up diff.

The distinction between SQLite as the control-plane authority and SQLite pages as the container for every canonical value is persuasive. I withdraw the monolithic-SQLite design objection as a merge blocker; the hybrid value layer is a coherent design when its publication and deletion protocol preserves the reviewed invariants.

Non-blocking follow-up: please amend #4071 so its physical-placement decision matches the implementation, and freeze the backup/restore contract, inline threshold, Worker/event-loop budget, and maintenance budgets before production writer activation.

Approved.

</details>

<details>
<summary>中文</summary>

感谢详细回复和针对性修复。我重新审查了 `c10089339`，并复跑了相关 build 与测试。

五项 finding 均已闭合：

- Read facade 已绑定 Session；
- 物理 placement 已收回到 authority，原 mixed-representation 复现现在会重新发布并校验缺失的 managed value，再返回成功；
- 两条 publication 路径都建立了必要的 durability barrier；
- deterministic staging 已进入 durable intent 与恢复路径；
- pending file bytes 和删除 continuation 会持续计入 accounting，直到持久清理完成。

`@maka/core` 与 `@maka/storage` build 通过，定向测试 31/31 通过。follow-up diff 中没有发现新的 merge blocker。

关于架构，作者对“SQLite 作为 control-plane authority”与“SQLite pages 容纳全部 canonical value”所做的区分是成立的。我不再把 monolithic SQLite 作为合并前提；只要 publication/deletion protocol 保持本轮确认的不变量，hybrid value layer 是自洽的设计。

非阻断 follow-up：请同步修改 #4071，使其中的 physical-placement 决策与实现一致；并在生产 writer 激活前冻结 backup/restore 合同、inline threshold、Worker/event-loop budget 与维护预算。

Approved。

</details>
